programs/bpf_loader/src/syscalls.rs

```diff
@@ -934,901 +1150,1892 @@ impl<'a> SyscallObject<BpfError> for SyscallSha256<'a> {
+        *result = Ok(0)
     }
 }
-/// Get a EpochSchedule sysvar
-struct SyscallGetEpochScheduleSysvar<'a> {
-    invoke_context: Rc<RefCell<&'a mut dyn InvokeContext>>,
+
+pub struct SyscallBigNumFromU32<'a> {
+    cost: u64,
+    compute_meter: Rc<RefCell<dyn ComputeMeter>>,
     loader_id: &'a Pubkey,
 }
-impl<'a> SyscallObject<BpfError> for SyscallGetEpochScheduleSysvar<'a> {
+impl<'a> SyscallObject<BpfError> for SyscallBigNumFromU32<'a> {
     fn call(
         &mut self,
-        var_addr: u64,
-        _arg2: u64,
+        bn_addr: u64,
+        u32_val: u64,
         _arg3: u64,
         _arg4: u64,
         _arg5: u64,
         memory_mapping: &MemoryMapping,
         result: &mut Result<u64, EbpfError<BpfError>>,
     ) {
-        *result = get_sysvar::<EpochSchedule>(
-            &sysvar::epoch_schedule::id(),
-            var_addr,
-            self.loader_id,
-            memory_mapping,
-            self.invoke_context.clone(),
+        let big_number = question_mark!(
+            translate_type_mut::<u64>(memory_mapping, bn_addr, self.loader_id, true),
+            result
         );
+        let bbox = Box::new(BigNum::from_u32(u32_val as u32).unwrap());
+        let bytes = bbox.num_bytes() as f64;
+        let rwptr = Box::into_raw(bbox);
+        let bignum_ptr = rwptr as u64;
+        *big_number = bignum_ptr;
+        question_mark!(
+            self.compute_meter
+                .consume(self.cost + calc_bignum_cost!(bytes)),
+            result
+        );
+        *result = Ok(0)
     }
 }
-/// Get a Fees sysvar
-struct SyscallGetFeesSysvar<'a> {
-    invoke_context: Rc<RefCell<&'a mut dyn InvokeContext>>,
+pub struct SyscallBigNumFromBytes<'a> {
+    cost: u64,
+    compute_meter: Rc<RefCell<dyn ComputeMeter>>,
     loader_id: &'a Pubkey,
 }
-impl<'a> SyscallObject<BpfError> for SyscallGetFeesSysvar<'a> {
+impl<'a> SyscallObject<BpfError> for SyscallBigNumFromBytes<'a> {
     fn call(
         &mut self,
-        var_addr: u64,
-        _arg2: u64,
-        _arg3: u64,
+        bn_addr: u64,
+        bytes_addr: u64,
+        bytes_len: u64,
         _arg4: u64,
         _arg5: u64,
         memory_mapping: &MemoryMapping,
         result: &mut Result<u64, EbpfError<BpfError>>,
     ) {
-        *result = get_sysvar::<Fees>(
-            &sysvar::fees::id(),
-            var_addr,
-            self.loader_id,
-            memory_mapping,
-            self.invoke_context.clone(),
+        let big_number = question_mark!(
+            translate_type_mut::<u64>(memory_mapping, bn_addr, self.loader_id, true),
+            result
+        );
+        let byte_slice = question_mark!(
+            translate_slice::<u8>(memory_mapping, bytes_addr, bytes_len, self.loader_id, true),
+            result
+        );
+        let bytes: f64 = byte_slice.len() as f64;
+        let bbox = Box::new(BigNum::from_slice(byte_slice).unwrap());
+        let rwptr = Box::into_raw(bbox);
```

Comment:
This screems memory leak, is it possible to create a container in program space (above the syscall) and just have the syscall fill in the result?

Reply:
Maybe pass back up a byte array and then call `from_slice()`?

Reply:
May need clarity here.

Reply:
The model here looks like it is allocating a bignum type in the syscall and then passing back a raw pointer, which in program space is converted into a bignum type.  This means the allocation is occurring in the runtime and relying on the program to trigger the drop.  A malicious program could skip the drop resulting in a memory leak in the runtime.

Reply:
Could you use `from_bytes` and `to_bytes` to pass a byte array between the two.  That does lead to local allocations on both ends which isn't ideal but better then a possible memory leak.

Reply:
@jackcmay - Wanting to grok the suggestion but haven't been working in rust for too long so bear with me...

1. Are you suggesting that instead of boxing, we pass arrays back and forth to syscalls and instantiate the openSSL BigNum in the syscall to perform the operations?
2. As the size of the arrays are variable, is there a way to pass mutable pointers to Vec<u8>, `translate` them to Vec<u8> in the syscall and push results? I'm having a darned time trying to figure this out.

Sorry for the n00b questions...

Reply:
@jackcmay So, remember I am a Rust n00b:
1. Instead of boxing and raw-pointers, send and receive byte arrays to/from syscalls and in the syscalls instantiate the BigNum's?
2. Is there a way to pass in a Vec<u8> and `translate` it to type in a syscall? I've had trouble and can't seem to make it work.
3. The reason for the Vec is that BigNum are ultimately variable length and can get quite large so knowing fixed buffers a priori would be daunting.